File: centsible-ui/app/services/user/user-service.ts
import type {AxiosInstance} from "axios";
import type {UserDto} from "~/models/user/user-dto";
import type {NotificationSettings} from "~/models/notification/notification-settings";
import type {Currency} from "~/models/budget-account/currency";
import {postMultipart, validateRequest} from "~/composables/use-api";

export function useUserService(api: AxiosInstance) {
  async function fetchMyself(): Promise<UserDto> {
    const response = await api.get<UserDto>('/users/myself');
    return validateRequest(response);
  }

  async function updateProfile(profile: { firstName: string, lastName: string, email: string }): Promise<UserDto> {
    const response = await api.put<UserDto>('/users/profile', profile);
    return validateRequest(response);
  }

  async function updateProfilePicture(file: File): Promise<UserDto> {
    return postMultipart<UserDto>(api, '/users/profile/picture', {file});
  }

  async function updateLocale(locale: string): Promise<UserDto> {
    const response = await api.put<UserDto>('/users/locale', {locale});
    return validateRequest(response);
  }

  async function updateDefaultCurrency(defaultCurrency: Currency): Promise<UserDto> {
    const response = await api.put<UserDto>('/users/default-currency', {defaultCurrency});
    return validateRequest(response);
  }

  async function fetchNotificationSettings(): Promise<NotificationSettings> {
    const response = await api.get<NotificationSettings>('/users/notification-settings');
    return validateRequest(response);
  }

  async function updateNotificationSettings(settings: NotificationSettings): Promise<NotificationSettings> {
    const response = await api.put<NotificationSettings>('/users/notification-settings', settings);
    return validateRequest(response);
  }

  return {
    fetchMyself,
    updateProfile,
    updateProfilePicture,
    updateLocale,
    updateDefaultCurrency,
    fetchNotificationSettings,
    updateNotificationSettings,
  }
}

File: centsible-core/src/main/kotlin/beer/thierry/centsible/core/services/users/UserService.kt
package beer.thierry.centsible.core.services.users

import beer.thierry.centsible.api.exceptions.LocalizedException
import beer.thierry.centsible.api.model.notification.NotificationSettingsDTO
import beer.thierry.centsible.api.model.user.ProfileUpdateDTO
import beer.thierry.centsible.api.model.user.User
import beer.thierry.centsible.api.model.user.UserDTO
import beer.thierry.centsible.api.repository.IUserRepository
import beer.thierry.centsible.api.services.users.IUserService
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import java.util.*

@Service
class UserService(private val userRepository: IUserRepository) : IUserService {

    private val log = LoggerFactory.getLogger(UserService::class.java)

    override fun fetchUserByUsername(username: String): UserDTO {
        val user = userRepository.findUserByUsername(username)
            ?: throw LocalizedException.NotFound("error.user.notFoundByUsername", username)
        return mapToDTO(user)
    }

    override fun userExists(id: UUID): Boolean = userRepository.findUserById(id) != null

    override fun updateUserProfile(userId: UUID, profile: ProfileUpdateDTO): UserDTO {
        val dto = persistProfile(
            userId = userId,
            failureKey = "error.user.profileUpdateFailed",
        ) { existing ->
            existing.copy(
                firstName = profile.firstName,
                lastName = profile.lastName,
                email = profile.email,
            )
        }
        log.info("Updated user profile userId={}", userId)
        return dto
    }

    override fun updateProfilePicture(userId: UUID, profilePicture: String?): UserDTO {
        val dto = persistProfile(
            userId = userId,
            failureKey = "error.user.pictureUpdateFailed",
        ) { existing ->
            existing.copy(profilePicture = profilePicture)
        }
        log.info("Updated user profile picture userId={} cleared={}", userId, profilePicture == null)
        return dto
    }

    override fun updateUserLocale(userId: UUID, locale: String): UserDTO {
        val updated = userRepository.updateUserLocale(userId, locale)
            ?: throw LocalizedException.NotFound("error.user.notFound")
        log.info("Updated user locale userId={} locale={}", userId, locale)
        return mapToDTO(updated)
    }

    override fun fetchNotificationSettings(userId: UUID): NotificationSettingsDTO =
        userRepository.fetchNotificationSettings(userId)

    override fun updateNotificationSettings(
        userId: UUID,
        settings: NotificationSettingsDTO,
    ): NotificationSettingsDTO {
        val updated = userRepository.updateNotificationSettings(userId, settings)
        log.info("Updated notification settings userId={}", userId)
        return updated
    }

    private fun persistProfile(userId: UUID, failureKey: String, mutate: (User) -> User): UserDTO {
        val existing = userRepository.findUserById(userId)
            ?: throw LocalizedException.NotFound("error.user.notFound")
        val patched = mutate(existing)
        val updated = userRepository.updateUserProfile(
            id = userId,
            firstName = patched.firstName,
            lastName = patched.lastName,
            email = patched.email,
            profilePicture = patched.profilePicture,
        ) ?: throw LocalizedException.InternalError(failureKey)
        return mapToDTO(updated)
    }

    private fun mapToDTO(user: User): UserDTO = UserDTO(
        id = user.id,
        username = user.username,
        email = user.email,
        firstName = user.firstName,
        lastName = user.lastName,
        name = "${user.firstName} ${user.lastName}",
        profilePicture = user.profilePicture,
        locale = user.locale,
    )
}
